extensions/typescript-language-features/src/typescriptServiceClient.ts

```diff
@@ -973,6 +975,45 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 			case EventName.projectLoadingFinish:
 				this.loadingIndicator.finishedLoadingProject((event as Proto.ProjectLoadingFinishEvent).body.projectName);
 				break;
+
+			case EventName.createDirectoryWatcher:
+				this.createFileSystemWatcher(event.body.id, new vscode.RelativePattern(vscode.Uri.file(event.body.path), event.body.recursive ? '**' : '*'));
```

Comment:
Note that this should work fine for non-recursive watchers (which are always preferred) but has some limitations for recursive watchers if the path is inside the workspace. Please review: https://code.visualstudio.com/api/references/vscode-api#workspace.createFileSystemWatcher

Reply:
I think I would need some help in fixing this part then. I think we would want to watch the folder if its workspace folder or outside of workspace otherwise watch workspace and only redirect to tsserver when events it is interested in are generated (inside the sub path area)
Does this need to handle more workspace folders added after watching has started etc. I am not sure how it works with all that but if someone more expert can take over that would be great.

Reply:
@sheetalkamat 

> I think I would need some help in fixing this part then.

Just to clarify: you are free to `createFileSystemWatcher` on any path from your extension, both recursive and non-recursive. However, one limitation today is that recursive watching inside the workspace is superseded by the watcher that VS Code itself sets up when you open the workspace. That means, `files.watcherExclude` setting is applied and by default `**/node_modules/*/**` will be ignored. So if you need to watch recursively inside `node_modules` folders from your extension, we need to think how to support that, either by VS Code making this possible or you changing to non-recursive watching if possible. So please let me know if that is the case.

> Does this need to handle more workspace folders added after watching has started etc

Both recursive and non-recursive file watching can handle the case of new folders being added. But if your extension needs to create new file system watchers for new workspaces added, then you need to handle that case from your extension by listening to respective workspace folder change events. Maybe I misunderstood the question, can you rephrase?

Reply:
We do need to watch node_modules and we watch them recursively either to watch when program references files from it (we prefer that over watching individual files) and also to ensure that resolutions can be updated eg when npm install happens.

Reply:
I am not sure I fully understand. Can you give examples of file changes in `node_modules` folder that you are interested in specifically?